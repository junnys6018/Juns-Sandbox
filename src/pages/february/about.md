Generative art is the process of creating art through the use of some autonomous process. In this sandbox we create generative art using cellular automata.

### The math

A cellular automata consists of a grid of cells, each cell can be in a finite number of states. We denote the set of possible states with $S$. The state of the cells have some initial given state at $t=0$, and is evolved over time by an update function $f:S^{n}\to S$. At each cell in our grid we look at $n$ neighbors surrounding the cell and compute the new state of the cell as a function of the state of its $n$ neighbors. How we pick the neighbors is up to us, however two types of neighborhoods are commonly used. The first is called the von Neumann neighborhood, here $n=5$ and we take the cell of interest and its four orthogonally adjacent cells as the neighborhood. The second is called the Moore neighborhood, where $n=9$ and we take the cells in a $3\times3$ grid surrounding the cell of interest as the neighborhood.

In this demo we use a 300 by 200 grid of cells with periodic boundary conditions.

### Stepping Stone Cellular Automata

In a stepping stone cellular automata we use a von Neumann neighborhood. The state of each cell is an $(r, g, b)$ color triple, so $S=\{0,1,...,255\}^3$. The update function is given by:

$$
f(c, u, d, l, r) = 
\begin{cases}
c, &\text{if } r < a \\
u, &\text{if } a \leq r < b \\
d, &\text{if } b \leq r < c \\
r, &\text{if } c \leq r < d \\
l, &\text{if } d \leq r
\end{cases}
$$

Where $c$ is the center cell of the von Neumann neighborhood, i.e. the cell we are updating. $u$, $d$, $l$, $r$ denotes the cells above, below, left and right of the cell of interest respectively. $r$ is a real number uniformly sampled from $[0, 1]$ and $a<b<c<d$ is a partition of $[0,1]$. In this implementation we take $a=0.5$, $b=0.625$, $c=0.75$, $d=0.875$, in other words, a cell has a 50% chance of not changing color, if a cell does change color, it is equally likely to change to any of its neighbor's colors.

### Cyclic Cellular Automata

A cyclic cellular automata is slightly more complicated, again we use a von Neumann neighborhood. This time state of each cell is an element of $S=\{1,...,k\}$, in our implementation we have $k=14$. The update function is:

$$
f(c, u, d, l, r) = 
\begin{cases}
\alpha, &\text{if } (u=\alpha) \vee (d=\alpha) \vee (l=\alpha) \vee (r=\alpha) \\
c,      &\text{otherwise}
\end{cases}
$$

Where $\alpha = (c + 1) \text{ mod } k$. In other words we increment the state of a cell modulo $k$ if any of its neighbors has a state equal to $\alpha$, otherwise the state of a cell does not change. We visualize the grid of cells by assigning a color to each of the $k$ possible states.

### References

[https://progur.com/2018/10/creating-generative-art-cellular.html](https://progur.com/2018/10/creating-generative-art-cellular.html)