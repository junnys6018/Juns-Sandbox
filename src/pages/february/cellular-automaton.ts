type color = [number, number, number];

export abstract class CellularAutomaton {
    abstract update(oldImage: ImageData, newImage: ImageData, x: number, y: number): void;

    init(_context: CanvasRenderingContext2D, _width: number, _height: number) { }
    finishFrame() { }

    // helper function
    get(image: ImageData) {
        return [
            image.width,
            image.height,
            [
                [1, 0],
                [image.width - 1, 0],
                [0, 1],
                [0, image.height - 1],
            ],
        ] as const;
    }
}

export class SteppingStone extends CellularAutomaton {
    init(context: CanvasRenderingContext2D, width: number, height: number) {
        const image = context.getImageData(0, 0, width, height);

        for (let y = 0; y < image.height; y++) {
            for (let x = 0; x < image.width; x++) {
                set(image, x, y, rand(255), rand(255), rand(255));
            }
        }

        context.putImageData(image, 0, 0);
    }

    update(oldImage: ImageData, newImage: ImageData, x: number, y: number): void {
        const [width, height, directions] = this.get(oldImage);

        if (Math.random() < 0.5) {
            // set to color of one of its neighbours
            const dir = directions[rand(4)];
            set(newImage, x, y, ...get(oldImage, (x + dir[0]) % width, (y + dir[1]) % height));
        } else {
            // keep color
            set(newImage, x, y, ...get(oldImage, x, y));
        }
    }
}

export class Cyclic extends CellularAutomaton {
    states: number[];
    newStates: number[];
    colors: color[];
    numStates: number;

    constructor(width: number, height: number, numStates: number) {
        super();
        this.numStates = numStates;
        this.states = Array<number>(width * height);
        this.newStates = Array<number>(width * height);

        this.colors = Array<color>(numStates);

        for (let i = 0; i < numStates; i++) {
            this.colors[i] = [rand(255), rand(255), rand(255)];
        }

        for (let i = 0; i < width * height; i++) {
            this.states[i] = rand(numStates);
        }
    }

    update(oldImage: ImageData, newImage: ImageData, x: number, y: number): void {
        const [width, height, directions] = this.get(oldImage);

        let currentState = this.states[y * width + x];
        const nextState = (currentState + 1) % this.numStates;
        let update = false;
        for (const d of directions) {
            const idx = ((y + d[1]) % height) * width + ((x + d[0]) % width);
            if (this.states[idx] === nextState) {
                update = true;
                break;
            }
        }

        if (update) {
            currentState = nextState;
        }

        this.newStates[y * width + x] = currentState;

        set(newImage, x, y, ...this.colors[currentState]);
    }

    finishFrame() {
        // update states
        this.states.splice(0, this.states.length, ...this.newStates);
    }
}

function rand(max: number) {
    return Math.floor(Math.random() * max);
}

function set(image: ImageData, x: number, y: number, r: number, g: number, b: number) {
    const idx = 4 * (y * image.width + x);
    image.data[idx + 0] = r;
    image.data[idx + 1] = g;
    image.data[idx + 2] = b;
    image.data[idx + 3] = 255;
}

function get(image: ImageData, x: number, y: number): color {
    const idx = 4 * (y * image.width + x);
    return [image.data[idx], image.data[idx + 1], image.data[idx + 2]];
}