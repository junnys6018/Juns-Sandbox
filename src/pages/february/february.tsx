import { RadioGroup } from '@headlessui/react';
import Footer from 'components/footer';
import Heading from 'components/heading';
import Navbar from 'components/navbar';
import Socials from 'components/socials';

import { Fragment, useCallback, useEffect, useRef, useState } from 'react';
import { FaPause, FaPlay, FaRedoAlt } from 'react-icons/fa';
import { CellularAutomaton, Cyclic, SteppingStone } from './cellular-automaton';
import aboutMarkdown from './about.md';
import { AsyncMarkdown } from 'markdown';

interface CellularAutomataComponentProps {
    width: number;
    height: number;
    children: (props: {
        paused: boolean;
        togglePaused: () => void;
        updateFunction: UpdateFunction;
        setUpdateFunction: (updateFunction: UpdateFunction) => void;
        downloadImage: () => void;
        reset: () => void;
    }) => React.ReactNode;
}

const canvasWidth = 300;
const canvasHeight = 200;

function CellularAutomataComponent(props: CellularAutomataComponentProps) {
    const canvasElement = useRef<HTMLCanvasElement>(null);

    const [paused, setPaused] = useState(false);
    const togglePaused = useCallback(() => setPaused(p => !p), []);

    const [updateFunction, _setUpdateFunction] = useState<UpdateFunction>('Stepping Stone');
    const [automaton, setAutomaton] = useState<CellularAutomaton>(new SteppingStone());

    const setUpdateFunction = useCallback((updateFunction: UpdateFunction) => {
        _setUpdateFunction(updateFunction);
        switch (updateFunction) {
            case 'Cyclic':
                setAutomaton(new Cyclic(canvasWidth, canvasHeight, 14));
                break;
            case 'Stepping Stone':
                setAutomaton(new SteppingStone());
                break;
        }
    }, []);

    const downloadImage = useCallback(() => {
        const canvas = canvasElement.current;
        if (!canvas) {
            return;
        }

        const link = document.createElement('a');
        link.download = 'generative-art.png';
        link.href = canvas.toDataURL();
        link.click();
    }, []);

    // call init on automaton every time it has changed
    const reset = useCallback(() => {
        const canvasContext = canvasElement.current?.getContext('2d');
        if (!canvasContext) {
            return;
        }
        automaton.init(canvasContext, props.width, props.height);
    }, [automaton, props.height, props.width]);

    useEffect(reset, [reset]);

    // animate the canvas
    useEffect(() => {
        const canvasContext = canvasElement.current?.getContext('2d');
        if (!canvasContext) {
            return;
        }

        let rafId: number;

        const raf = () => {
            rafId = requestAnimationFrame(raf);

            if (paused) {
                return;
            }

            const oldImage = canvasContext.getImageData(0, 0, props.width, props.height);
            const newImage = canvasContext.createImageData(props.width, props.height);

            for (let y = 0; y < props.height; y++) {
                for (let x = 0; x < props.width; x++) {
                    automaton.update(oldImage, newImage, x, y);
                }
            }
            automaton.finishFrame();
            canvasContext.putImageData(newImage, 0, 0);
        };

        rafId = requestAnimationFrame(raf);

        return () => {
            cancelAnimationFrame(rafId);
        };
    }, [automaton, paused, props.height, props.width]);

    return (
        <Fragment>
            <canvas
                className="flex-grow pixelated max-w-full"
                width={props.width}
                height={props.height}
                ref={canvasElement}
            ></canvas>
            {props.children({ paused, togglePaused, updateFunction, setUpdateFunction, downloadImage, reset })}
        </Fragment>
    );
}

interface SettingsPanelProps {
    paused: boolean;
    togglePaused: () => void;
    updateFunction: UpdateFunction;
    setUpdateFunction: (updateFunction: UpdateFunction) => void;
    downloadImage: () => void;
    reset: () => void;
}

type UpdateFunction = 'Stepping Stone' | 'Cyclic';

function UpdateFunctionOption(props: { value: UpdateFunction }) {
    return (
        <RadioGroup.Option value={props.value} className="flex flex-row items-center gap-1.5 cursor-pointer">
            {({ checked }) => (
                <Fragment>
                    <div className="bg-neutral-300 rounded-full w-4 h-4 flex items-center justify-center">
                        {checked && (
                            <div className="w-2.5 h-2.5 bg-pink-500 rounded-full shadow-[0_0_4px_#ec4899]"></div>
                        )}
                    </div>
                    <RadioGroup.Label className="font-medium">{props.value}</RadioGroup.Label>
                </Fragment>
            )}
        </RadioGroup.Option>
    );
}

function SettingsPanel(props: SettingsPanelProps) {
    return (
        <div className="flex flex-col items-start drop-shadow-lg bg-white rounded-3xl p-8 flex-shrink-0 w-full xl:w-[470px]">
            <div className="flex flex-row items-center gap-10">
                <button
                    className="bg-pink-500 hover:bg-pink-600 rounded-full w-24 h-24 flex items-center justify-center"
                    onClick={props.togglePaused}
                >
                    {props.paused ? (
                        <FaPlay className="text-neutral-50 w-8 h-8" />
                    ) : (
                        <FaPause className="text-neutral-50 w-8 h-8" />
                    )}
                </button>

                <button
                    onClick={props.reset}
                    className="bg-neutral-200 hover:bg-neutral-300 rounded-full w-16 h-16 flex items-center justify-center"
                >
                    <FaRedoAlt className="text-neutral-900 w-6 h-6" />
                </button>
            </div>

            <h3 className="font-medium text-neutral-900 mb-2.5 mt-6">Update Function</h3>
            <RadioGroup value={props.updateFunction} onChange={props.setUpdateFunction} className="mb-6">
                <UpdateFunctionOption value="Stepping Stone" />
                <UpdateFunctionOption value="Cyclic" />
            </RadioGroup>

            <button
                onClick={props.downloadImage}
                className="flex items-center justify-center mt-auto font-medium bg-pink-500 hover:bg-pink-600 rounded-lg text-neutral-50 py-2 px-3"
            >
                Download Image
            </button>
        </div>
    );
}

export default function February() {
    return (
        <div className="flex flex-col min-h-screen">
            <Navbar />
            <div className="container mb-8">
                <Heading month={2} />

                <div className="flex flex-row flex-wrap mt-6 gap-x-10 gap-y-6">
                    <CellularAutomataComponent width={canvasWidth} height={canvasHeight}>
                        {props => <SettingsPanel {...props} />}
                    </CellularAutomataComponent>
                </div>
                <h1 className="font-semibold text-4xl text-pink-500 my-8">About this demo</h1>
                <AsyncMarkdown className="max-w-none" src={aboutMarkdown} />
            </div>
            <Footer />
            <Socials />
        </div>
    );
}
