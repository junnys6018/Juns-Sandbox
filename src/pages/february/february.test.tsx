import { render, screen } from '@testing-library/react';
import { SettingsPanel, SettingsPanelProps } from './february';
import userEvent from '@testing-library/user-event';

test('<SettingsPanel /> snapshot', () => {
    const props: SettingsPanelProps = {
        paused: false,
        togglePaused: jest.fn(),
        updateFunction: 'Cyclic',
        setUpdateFunction: jest.fn(),
        downloadImage: jest.fn(),
        reset: jest.fn(),
    };
    const { asFragment } = render(<SettingsPanel {...props} />);
    expect(asFragment()).toMatchSnapshot();
});

test('<SettingsPanel /> user interaction', () => {
    const props: SettingsPanelProps = {
        paused: false,
        togglePaused: jest.fn(),
        updateFunction: 'Cyclic',
        setUpdateFunction: jest.fn(),
        downloadImage: jest.fn(),
        reset: jest.fn(),
    };

    render(<SettingsPanel {...props} />);

    // update function radio buttons
    const steppingStoneLabel = screen.getByLabelText('Stepping Stone');
    userEvent.click(steppingStoneLabel);
    expect(props.setUpdateFunction).toHaveBeenLastCalledWith('Stepping Stone');

    // download image button
    const downloadImageButton = screen.getByText('Download Image');
    userEvent.click(downloadImageButton);
    expect(props.downloadImage).toHaveBeenCalledTimes(1);
});
