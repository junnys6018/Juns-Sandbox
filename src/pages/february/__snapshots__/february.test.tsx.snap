// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsPanel /> snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col items-start drop-shadow-lg bg-white rounded-3xl p-8 flex-shrink-0 w-full xl:w-[470px]"
  >
    <div
      class="flex flex-row items-center gap-10"
    >
      <button
        class="bg-pink-500 hover:bg-pink-600 rounded-full w-24 h-24 flex items-center justify-center"
      >
        <svg
          class="text-neutral-50 w-8 h-8"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M144 479H48c-26.5 0-48-21.5-48-48V79c0-26.5 21.5-48 48-48h96c26.5 0 48 21.5 48 48v352c0 26.5-21.5 48-48 48zm304-48V79c0-26.5-21.5-48-48-48h-96c-26.5 0-48 21.5-48 48v352c0 26.5 21.5 48 48 48h96c26.5 0 48-21.5 48-48z"
          />
        </svg>
      </button>
      <button
        class="bg-neutral-200 hover:bg-neutral-300 rounded-full w-16 h-16 flex items-center justify-center"
      >
        <svg
          class="text-neutral-900 w-6 h-6"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256.455 8c66.269.119 126.437 26.233 170.859 68.685l35.715-35.715C478.149 25.851 504 36.559 504 57.941V192c0 13.255-10.745 24-24 24H345.941c-21.382 0-32.09-25.851-16.971-40.971l41.75-41.75c-30.864-28.899-70.801-44.907-113.23-45.273-92.398-.798-170.283 73.977-169.484 169.442C88.764 348.009 162.184 424 256 424c41.127 0 79.997-14.678 110.629-41.556 4.743-4.161 11.906-3.908 16.368.553l39.662 39.662c4.872 4.872 4.631 12.815-.482 17.433C378.202 479.813 319.926 504 256 504 119.034 504 8.001 392.967 8 256.002 7.999 119.193 119.646 7.755 256.455 8z"
          />
        </svg>
      </button>
    </div>
    <h3
      class="font-medium text-neutral-900 mb-2.5 mt-6"
    >
      Update Function
    </h3>
    <div
      class="mb-6"
      id="headlessui-radiogroup-7"
      role="radiogroup"
    >
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-1"
        class="flex flex-row items-center gap-1.5 cursor-pointer"
        id="headlessui-radiogroup-option-2"
        role="radio"
        tabindex="-1"
      >
        <div
          class="bg-neutral-300 rounded-full w-4 h-4 flex items-center justify-center"
        />
        <label
          class="font-medium"
          id="headlessui-label-1"
        >
          Stepping Stone
        </label>
      </div>
      <div
        aria-checked="true"
        aria-labelledby="headlessui-label-3"
        class="flex flex-row items-center gap-1.5 cursor-pointer"
        id="headlessui-radiogroup-option-4"
        role="radio"
        tabindex="0"
      >
        <div
          class="bg-neutral-300 rounded-full w-4 h-4 flex items-center justify-center"
        >
          <div
            class="w-2.5 h-2.5 bg-pink-500 rounded-full shadow-[0_0_4px_#ec4899]"
          />
        </div>
        <label
          class="font-medium"
          id="headlessui-label-3"
        >
          Cyclic
        </label>
      </div>
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-5"
        class="flex flex-row items-center gap-1.5 cursor-pointer"
        id="headlessui-radiogroup-option-6"
        role="radio"
        tabindex="-1"
      >
        <div
          class="bg-neutral-300 rounded-full w-4 h-4 flex items-center justify-center"
        />
        <label
          class="font-medium"
          id="headlessui-label-5"
        >
          Game of Life
        </label>
      </div>
    </div>
    <button
      class="flex items-center justify-center mt-auto font-medium bg-pink-500 hover:bg-pink-600 rounded-lg text-neutral-50 py-2 px-3"
    >
      Download Image
    </button>
  </div>
</DocumentFragment>
`;
